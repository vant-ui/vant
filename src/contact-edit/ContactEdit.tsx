import { watch, reactive, PropType, defineComponent } from 'vue';

// Utils
import { createNamespace } from '../utils';
import { isMobile } from '../utils/validate/mobile';

// Components
import Cell from '../cell';
import Form from '../form';
import Field from '../field';
import Button from '../button';
import Dialog from '../dialog';
import Switch from '../switch';

const [name, bem, t] = createNamespace('contact-edit');

export type ContactEditInfo = {
  tel: string;
  name: string;
  isDefault?: boolean;
};

const DEFAULT_CONTACT: ContactEditInfo = {
  tel: '',
  name: '',
};

export default defineComponent({
  name,

  props: {
    isEdit: Boolean,
    isSaving: Boolean,
    isDeleting: Boolean,
    showSetDefault: Boolean,
    setDefaultLabel: String,
    contactInfo: {
      type: Object as PropType<ContactEditInfo>,
      default: () => ({ ...DEFAULT_CONTACT }),
    },
    telValidator: {
      type: Function as PropType<(val: string) => boolean>,
      default: isMobile,
    },
  },

  emits: ['save', 'delete', 'change-default'],

  setup(props, { emit }) {
    const contact = reactive({
      ...DEFAULT_CONTACT,
      ...props.contactInfo,
    });

    const onSave = () => {
      if (!props.isSaving) {
        emit('save', contact);
      }
    };

    const onDelete = () => {
      Dialog.confirm({
        title: t('confirmDelete'),
      }).then(() => emit('delete', contact));
    };

    const renderButtons = () => (
      <div class={bem('buttons')}>
        <Button
          block
          round
          type="danger"
          text={t('save')}
          loading={props.isSaving}
          nativeType="submit"
        />
        {props.isEdit && (
          <Button
            block
            round
            text={t('delete')}
            loading={props.isDeleting}
            onClick={onDelete}
          />
        )}
      </div>
    );

    const renderSwitch = () => (
      <Switch
        v-model={contact.isDefault}
        size={24}
        onChange={(checked: boolean) => emit('change-default', checked)}
      />
    );

    const renderSetDefault = () => {
      if (props.showSetDefault) {
        return (
          <Cell
            v-slots={{ 'right-icon': renderSwitch }}
            title={props.setDefaultLabel}
            class={bem('switch-cell')}
            border={false}
          />
        );
      }
    };

    watch(
      () => props.contactInfo,
      (value) => {
        Object.assign(contact, DEFAULT_CONTACT, value);
      }
    );

    return () => (
      <Form class={bem()} onSubmit={onSave}>
        <div class={bem('fields')}>
          <Field
            v-model={contact.name}
            clearable
            label={t('name')}
            rules={[{ required: true, message: t('nameInvalid') }]}
            maxlength="30"
            placeholder={t('nameEmpty')}
          />
          <Field
            v-model={contact.tel}
            clearable
            type="tel"
            label={t('tel')}
            rules={[
              { validator: props.telValidator, message: t('telInvalid') },
            ]}
            placeholder={t('telEmpty')}
          />
        </div>
        {renderSetDefault()}
        {renderButtons()}
      </Form>
    );
  },
});
