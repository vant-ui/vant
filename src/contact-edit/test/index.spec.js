import ContactEdit from '..';
import { mount, later } from '../../../test';

const contactInfo = {
  name: 'foo',
  tel: '13000000000',
};

test('should validate contact name before submitting form', async () => {
  const wrapper = mount(ContactEdit, {
    props: {
      contactInfo: {
        name: '',
        tel: '',
      },
    },
  });

  const button = wrapper.find('.van-button');
  button.trigger('click');
  await later();
  expect(wrapper.find('.van-field__error-message').html()).toMatchSnapshot();

  const fields = wrapper.findAll('.van-field__control');

  fields[0].trigger('focus');
  await later();
  expect(wrapper.find('.van-field__error-message').exists()).toBeFalsy();
});

test('should validate contact tel before submitting form', async () => {
  const wrapper = mount(ContactEdit, {
    props: {
      contactInfo: {
        name: 'name',
        tel: '',
      },
    },
  });

  const button = wrapper.find('.van-button');
  button.trigger('click');
  await later();
  expect(wrapper.find('.van-field__error-message').html()).toMatchSnapshot();

  const fields = wrapper.findAll('.van-field__control');

  fields[1].trigger('focus');
  await later();
  expect(wrapper.find('.van-field__error-message').exists()).toBeFalsy();
});

test('should emit save event after submitting form', async () => {
  const wrapper = mount(ContactEdit, {
    props: {
      contactInfo,
    },
  });

  const button = wrapper.find('.van-button');
  button.trigger('click');
  await later();
  expect(wrapper.emitted('save')[0][0]).toEqual(contactInfo);
});

test('should watch contact info', async () => {
  const wrapper = mount(ContactEdit);
  const button = wrapper.find('.van-button');

  wrapper.setProps({ contactInfo });
  await later();
  button.trigger('click');
  expect(wrapper.emitted('save')[0][0]).toEqual(contactInfo);
});

test('should allow deleting contact', async () => {
  const wrapper = mount(ContactEdit, {
    props: {
      isEdit: true,
    },
  });

  const deleteButton = wrapper.findAll('.van-button')[1];
  deleteButton.trigger('click');

  await later();
  document.querySelector('.van-dialog__confirm').click();

  await later();
  expect(wrapper.emitted('delete')).toBeTruthy();
});
