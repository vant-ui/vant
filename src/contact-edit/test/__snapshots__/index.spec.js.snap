// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should validate contact name before submitting form 1`] = `
<div class="van-field__error-message">
  请输入正确的姓名
</div>
`;

exports[`should validate contact tel before submitting form 1`] = `
<div class="van-field__error-message">
  请输入正确的手机号
</div>
`;
